src/test/java/testsuite/failover/FailoverIntegrationTest.java

```diff
@@ -928,140 +984,129 @@ private void assertFirstQueryThrows(Statement stmt, String expectedSQLErrorCode)
 
-  @AfterEach
-  private void reviveInstancesAndCloseTestConnection() throws SQLException {
-    reviveAllInstances();
-    testConnection.close();
+    makeSureInstancesUp(instanceIDs, false);
   }
 
-  private void reviveAllInstances() {
-    this.log.logDebug("Revive all crashed instances in the test and wait until they are up.");
-    instanceCrasherMap.forEach(
-        (instanceId, instanceCrasher) -> {
-          try {
-            stopCrashingInstanceAndWaitUntilUp(instanceId);
-          } catch (InterruptedException ex) {
-            this.log.logDebug("Exception occured while trying to stop crashing an instance.");
-          }
-        });
+  protected void startCrashingInstances(String... instances) {
+    instancesToCrash.addAll(Arrays.asList(instances));
   }
 
-  private void stopCrashingInstanceAndWaitUntilUp(String instanceId)
-      throws InterruptedException {
-    this.log.logDebug("Stop crashing " + instanceId + ".");
-    CrashInstanceRunnable instanceCrasher = instanceCrasherMap.get(instanceId);
-    instanceCrasher.stopCrashingInstance();
-    waitUntilInstanceIsUp(instanceId);
+  protected void stopCrashingInstances(String... instances) {
+    instancesToCrash.removeAll(Arrays.asList(instances));
   }
 
-  /**
-   * Start crashing the specified instance and wait until its inaccessible.
-   * */
-  private void startCrashingInstanceAndWaitUntilDown(String instanceId)
-      throws InterruptedException {
-    this.log.logDebug("Start crashing " + instanceId + ".");
-    CrashInstanceRunnable instanceCrasher = instanceCrasherMap.get(instanceId);
-    instanceCrasher.startCrashingInstance();
-    Thread instanceCrasherThread = new Thread(instanceCrasher);
-    instanceCrasherThread.start();
-    waitUntilInstanceIsDown(instanceId);
+  protected void makeSureInstancesUp(String... instances) throws InterruptedException {
+    makeSureInstancesUp(instances, true);
   }
 
-  /**
-   * Runnable class implementation that is used to crash an instance.
-   * */
-  public class CrashInstanceRunnable implements Runnable {
-    static final String CRASH_QUERY = "ALTER SYSTEM CRASH INSTANCE";
-    private final String instanceId;
-    private boolean keepCrashingInstance = false;
-
-    CrashInstanceRunnable(String instanceId) {
-      System.out.println("create runnable for " + instanceId);
-      this.instanceId = instanceId;
+  protected void makeSureInstancesUp(String[] instances, boolean finalCheck) throws InterruptedException {
+    this.log.logDebug("Wait until the following instances are up: \n" + String.join("\n", instances));
+    ExecutorService executorService = Executors.newFixedThreadPool(instances.length);
+    final HashSet<String> remainingInstances = new HashSet<String>();
+    remainingInstances.addAll(Arrays.asList(instances));
+
+    for (String id : instances) {
+      executorService.submit(() -> {
+        while (true) {
+          try (Connection conn = createCrashConnection(id)) {
```

Comment:
should we rename createCrashConnection because the implementation of the method itself doesn't have to do with crashing? And it seems like this is replacing createConnectionWithProxyDisabled; looks like createConnectionWithProxyDisabled isn't being called anymore. Maybe instead of creating this new method, just update the createConnectionWithProxyDisabled implementation with the createCrashConnection implementation?

Reply:
What'd be a better name for createCrashConnection ?

Reply:
Sure, I removed createCrashConnection and use createConnectionWithProxyDisabled